src/ess/dream/diagnostics.py

```diff
@@ -0,0 +1,228 @@
+        old_v = self._vertical_buttons.value
+        default_h, default_v = self._default_dims(new_dims)
+        new_h = old_h if old_h in new_dims else default_h
+        new_v = old_v if old_v in new_dims else default_v
+
+        h_buttons, v_buttons = self._make_buttons(new_dims, new_h, new_v)
+        self._horizontal_selector_box.children = [h_buttons]
+        self._vertical_selector_box.children = [v_buttons]
+
+    @staticmethod
+    def _default_dims(dims: tuple[str, ...]) -> tuple[str, str]:
+        return dims[math.ceil(len(dims) / 2)], dims[0]
+
+    @property
+    def value(self):
+        return {
+            'horizontal': self._horizontal_buttons.value,
+            'vertical': self._vertical_buttons.value,
+        }
+
+    def observe(self, handler, names: str | list[str], *args, **kwargs):
+        if names != 'value':
+            super().observe(handler, names, *args, **kwargs)
+        else:
+            self._value_observers.append(handler)
+
+    def _observe_values(self, change: dict) -> None:
+        clicked = change['owner']
+        other = (
+            self._vertical_buttons
+            if clicked is self._horizontal_buttons
+            else self._horizontal_buttons
+        )
+        if other.value == clicked.value:
+            other.value = change['old']
+        else:
+            # Only when dims are different to avoid unnecessary redraws
+            # and drawing with invalid dims.
+            for observer in self._value_observers:
+                observer(change)
+
+
+def _flat_voxel_figure(
+    data: sc.DataArray, horizontal_dim: str, vertical_dim: str
+) -> FigureLike:
+    # flat, h_coord, v_coord = _flatten_voxel_dims(data, horizontal_dim, vertical_dim)
+    kept_dims = {horizontal_dim, vertical_dim}
+
+    to_flatten = [dim for dim in data.dims if dim not in kept_dims]
+    n = len(to_flatten)
+    flatten_to_h = [horizontal_dim, *to_flatten[n // 2 :]]
+    flatten_to_v = [vertical_dim, *to_flatten[: n // 2]]
+
+    # Drop unused coordinates
+    aux = data.drop_coords(list(set(data.coords.keys()) - kept_dims))
+    reordered = aux.transpose(flatten_to_v + flatten_to_h)
+
+    h_coord = reordered.coords.pop(horizontal_dim)
+    v_coord = reordered.coords.pop(vertical_dim)
+
+    flat = reordered.flatten(flatten_to_v, to='vertical').flatten(
+        flatten_to_h, to='horizontal'
+    )
+    flat = flat.assign_coords(
+        {name: sc.arange(name, flat.sizes[name], unit=None) for name in flat.dims}
+    )
+
+    # This relies on the order of flatten_to_h/v
+    inner_volume_h = _product(data.sizes[d] for d in flatten_to_h[1:])
+    inner_volume_v = _product(data.sizes[d] for d in flatten_to_v[1:])
+    h_ticks = np.arange(0, flat.sizes['horizontal'], inner_volume_h)
+    v_ticks = np.arange(0, flat.sizes['vertical'], inner_volume_v)
+
+    h_labels = [str(value) for value in h_coord.values]
+    v_labels = [str(value) for value in v_coord.values]
+
+    fig = flat.plot(rasterized=True)
```

Comment:
I would probably make `rasterized` a kwarg for creating the viewer? (with `True` as default).